Feature: Global Path Parameters - for APIs that have the same base path parameter on every path

bbbates opened the pull request:
The Unifi API has a `siteId` path parameter that appears in (almost) all paths in the API. See: https://github.com/bbbates/pulumi-unifi-native/blob/81e014360d23ce5172b7e8d8f41a933bd5d0bef6/provider/cmd/pulumi-gen-unifi-native/openapi.yml#L12  (and in all the paths).

Ideally, users of the provider should not need to set this in each of their resources, as it would be quite noisy.

This feature is to allow a provider to encapsulate a map of global path parameters that can be defined outside of a resource (i.e. via configuration) and used when populating the set of path parameters for a request. The global path params are checked last when building the request path param map, so they can be overridden by resources if required. 

For example, this is how the current version of the Unifi provider configures this feature: https://github.com/bbbates/pulumi-unifi-native/blob/81e014360d23ce5172b7e8d8f41a933bd5d0bef6/provider/pkg/provider/provider.go#L152C1-L175C3 

```
func (p *unifiNativeProvider) OnConfigure(_ context.Context, req *pulumirpc.ConfigureRequest) (*pulumirpc.ConfigureResponse, error) {
	siteId, ok := req.GetVariables()["unifi-native:config:siteId"]
	if !ok {
		// Check if it's set as an env var.
		envVarNames := handler.GetSchemaSpec().Provider.InputProperties["siteId"].DefaultInfo.Environment
		...
	}

	logging.V(3).Infof("Configuring Site Id: %s", siteId)
	p.siteId = siteId
	if p.siteId != "" {
		handler.GetGlobalPathParams()["siteId"] = p.siteId
		handler.GetGlobalPathParams()["site_id"] = p.siteId
	}
```

Again, happy to discuss alternative solutions!

praneetloke reviewed (commented):
Thank you so much for tackling this! This PR fixes #23 that opened almost 3 years ago!

I added a suggestion for you to consider. Can you please take a look at it?

bbbates commented on rest/provider.go:
```diff
@@ -58,6 +58,10 @@ type Provider struct {
+
+	// Global path params for this provider - for path params that are fixed
+	// for a provider. Can be configured during the OnConfigure callback func
+	globalPathParams map[string]string
```
That would break backwards compatibility, wouldn't it? I tried my best to avoid that :upside_down_face:

bbbates commented:
@praneetloke How about this?

- Added a `GetGlobalParams` method to the `ProviderCallback` interface
- Added that "default" unimplemented type as well (the Template project will probably need to be updated to use that....)
- Removed the method that isn't used anymore

bbbates commented on callback/callback.go:
```diff
@@ -64,3 +69,61 @@ type ProviderCallback interface {
+
+type UnimplementedProviderCallback struct{}
+
+func (UnimplementedProviderCallback) GetAuthorizationHeader() string {
+	return ""
+}
```
They'll be forced to implement it because if they don't, their provider won't work :shrug:

bbbates replied:
I changed it anyway...

bbbates commented on callback/callback.go:
```diff
@@ -64,3 +69,61 @@ type ProviderCallback interface {
+}
+
+func (UnimplementedProviderCallback) OnConfigure(context.Context, *pulumirpc.ConfigureRequest) (*pulumirpc.ConfigureResponse, error) {
+	return nil, fmt.Errorf("OnConfigure not implemented")
```
Consistency is good

praneetloke replied:
I restored this comment thread again. I don't think this is resolved.

To clarify my original comment, I think these default implementations should _not_ return an error. That would force providers to have to implement these callbacks when they embed this `UnimplementedProviderCallback` struct in their provider interfaces, since the default implementation for all these callbacks return errors. Instead, these should be changed to match what is being done in the template repo today: https://github.com/cloudy-sky-software/pulumi-provider-template/blob/main/provider/pkg/provider/provider.go. Except for `OnConfigure`. I think by default, `OnConfigure` callback should simply return:

```go
func (UnimplementedProviderCallback) OnConfigure(_ context.Context, _ *pulumirpc.ConfigureRequest) (*pulumirpc.ConfigureResponse, error) {
	return &pulumirpc.ConfigureResponse{
		AcceptSecrets: true,
	}, nil
}
```
